[messages][messagegui] - fix music info handling

kidneyhex opened the pull request:
_**Closing this: The fix requires design choices, so I'll just log an issue and let someone else decide.**_ :P

bobrippling commented:
I'll hold off reviewing / testing until this is out of draft, but with it being a core app I'll tag @gfwilliams in the meanwhile

bobrippling commented:
Cool! For that last bit, you could ask in [discussions](https://github.com/orgs/espruino/discussions), I think there's a few iPhone users about

gfwilliams commented:
I'm really not convinced about requesting state (it feels like a hack to work around a bug somewhere) - I've put some more of my thoughts on this on https://codeberg.org/Freeyourgadget/Gadgetbridge/issues/4886#issuecomment-8622057

edit:

>  I ran into an app where GadgetBridge isn't sending musicinfo even on play/pause (libro.fm)

This is kind of what I mean - we should fix this in Gadgetbridge (or maybe it's a libro.fm issue?), and then it'll help everyone using it (even those without a Bangle.js).

kidneyhex commented:
I've updated this with logic to store the music in the `require('messages').music`, and then on kill event persist it into `messages.music.json`

Then `messagegui` uses `require('messages').getMusic()` to read the latest music info and there's a `_lastPlayed` flag to help expire old messages.

This still has a debug method in it to count number of times it writes to file--so not really ready for review/merge. Just here for info purposes for now.

gfwilliams commented:
> Just removing that return makes it write a ton

Are you sure? because I thought the messages app was supposed to be smart enough to only write to flash when apps were switched (but maybe I'm mis-remembering)

kidneyhex commented:
I think the messagegui listener writes to flash for every message if you don't have auto-open enabled.

https://github.com/espruino/BangleApps/blob/702db69c399382c3463b8f1babbb12d97b73611d/apps/messagegui/lib.js#L48

It only does the write on kill in the app.js itself.

gfwilliams commented:
Ok, thanks - I'm looking into this.

Can you see any reason why making messages do a delayed write for every message type wouldn't work? I'm setting it up so `require("messages").getMessages()` would still return the updated messages even if they weren't saved right away.

kidneyhex commented:
Yep, I think that'd work fine.

kidneyhex commented:
I was getting about 60% full of trash per day. I didn't dig in to figure out for sure why it was happening, but stopped as soon as I reverted my changes.




On Android there's a "setRating" method on the same class as the play/pause methods and ratings are stored in the metadata with artist/album. I've only seen Music Boss on Pebble use it though.

At first, I modified Gadgetbridge with new media commands to do it. But I've started adding adapters for apps that don't follow the standard, so I'm working on a companion app that uses intents. Eventually I'll probably publish the app and source so anyone with tasker/whatever can also use it.

I keep hoping someone will beat me to it, but so far there's always a little bit missing. Music Boss Deluxe has intents for setting ratings but doesn't have a way to read the current rating. Tasker lets you read the rating but doesn't let you set it. etc. etc.
